### ISSUE TITLE
Loading datacenter config contains a hidden nextTick even though all config loading is synchronous.

### ISSUE THREAD

**Issue by lxe** [uber/zero-config#31]
https://github.com/uber/zero-config/blob/master/index.js#L30-L34

```javascript
        // throw error async. this allows for breaking a 
        // circular dependency between config & logger.
        process.nextTick(function () {
            config.emit('error', err);
        });
```

All config loading is synchrnonous. `fetchConfig()` returns a result, and THEN throws an error, which is undesirable and confusing. For example:

```
test('Non-existent dc test', function assert(t) {
  t.throws(function shouldThrow() {
    var config = zc({
      dc: '/does/not/exist',
    });
  }, /MissingDatacenterFileError/);

  t.end();
});
```

^ This tests will fail, and THEN the test suite will crash.

**Comment by lxe** [comment]
I guess I can register on `error` event, but this just undoes the synchronous nature of config initialization.

**Comment by Raynos** [comment]
As mentioned; errors need to throw async because production. There are examples on how to write the tests.

**Comment by lxe** [comment]
> errors need to throw async because production.

I agree that If an error occurs during runtime, after the things are configured, it should of course log. If config is not initialized with all the required parameters, the process should crash. 

Things that are synchronous should stay synchronous. If async operations are being performed in the background, it should do `zeroConfig(..., function onConfigLoad(err, config) { ... });`. Or, following the existing EE pattern: `zeroConfig.on('load', function onConfigLoad(err, config) { ... })`

**Comment by lxe** [comment]
Why do other things throw in this case? Like:

https://github.com/uber/zero-config/blob/master/index.js#L15

**Comment by Raynos** [comment]
Because one is a programmer error and one is an operational error.

The programmer errors are thrown and the operational errors are emitted with the standard node event emitter interface
